### PR TITLE
KEYCLOAK-17665 Add query parameter support for Spring

### PR DESCRIPTION
**Pull request by dweber019**
Pull request to fix https://issues.redhat.com/browse/KEYCLOAK-17665

### CODE REVIEW

**Comment by dweber019** [comment]
Tried to run my newly created test in Intellij, doesn't work because of:
```
../keycloak/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authorization/PolicyEnforcer.java:35:41
java: package org.keycloak.authorization.client does not exist
```
Tried to build the whole keycloak project with `mvn clean install` as described in the docs. Error:
```
[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.097 s <<< FAILURE! - in org.keycloak.utils.StreamsUtilTest
[ERROR] testLimitOnClosingStream(org.keycloak.utils.StreamsUtilTest)  Time elapsed: 0.064 s  <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<3>
```
I did not change this?

Then I tried to do the same as in the github flows:
```
mvn install -nsu -B -DskipTestsuite -DskipExamples
```
Same error as before.

What should I do?

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java [hunk 1]
```diff
@@ -105,7 +106,13 @@ protected void commenceLoginRedirect(HttpServletRequest request, HttpServletResp
             // stored in a session. We'll store it in a cookie instead.
             response.addCookie(KeycloakCookieBasedRedirect.createCookieFromRedirectUrl(request.getRequestURI()));
         }
-        String contextAwareLoginUri = request.getContextPath() + loginUri;
+
+        String queryParameters = "";
+        if (!StringUtils.isEmpty(request.getQueryString())) {
+            queryParameters = "?" + request.getQueryString();
```
Comment (mhajas) [inline adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java:112]: I am not sure we want to add all parameters. Maybe we should use only those that can be consumed by Keycloak. WDYT?

File: adapters/oidc/spring-security/src/test/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPointTest.java [hunk 2]
```diff
@@ -78,6 +81,17 @@ public void testCommenceWithRedirect() throws Exception {
         assertEquals(KeycloakAuthenticationEntryPoint.DEFAULT_LOGIN_URI, response.getHeader("Location"));
     }
 
+    @Test
+    public void testCommenceWithRedirectAndQueryParameters() throws Exception {
+        configureBrowserRequest();
+        request.addParameter("prompt", "login");
```
Comment (mhajas) [inline adapters/oidc/spring-security/src/test/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPointTest.java:87]:
```suggestion
        request.setQueryString("prompt=login");
```

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you very much for the PR @dweber019! I added two comments in code.

I am sure why you are not able to run the test from Intellij, it works for me. The test is passing after the change I suggested in test comment. 

Maybe try `mvn clean install -DskipTests` from the keycloak-parent module. Then try to Reload maven project from Intellij and rebuild whole project.

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java (hunk 1 above)
Comment (dweber019) [reply adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java:112]: Currently the https://github.com/keycloak/keycloak/blob/cd342ad5714f15db1cc8b0cd55b788e6543c6dc8/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/OAuthRequestAuthenticator.java will check which query parameters are moved up to top url (not redirect uri), e.g. prompt=login. All other parameters are passed as is.
For us (my company) we define the language the application should use by setting the query param "lang=en" on the application url. If we don't strip this parameter we can check this in our Keycloak authenticators too.
In addition now it's the same behavior as in the JBoss adapter.

Comment (mhajas) [reply adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java:112]: You are right @dweber019, this makes sense. Thank you very much for further explanation.

**Comment by dweber019** [comment]
@mhajas can you help me regarding the following blocking reviews:
- Base testsuite (quarkus, group3), I think my PR isn't the issue here, right?
- Base testsuite (wildfly, group3), same here, right?
- As we accepted this solution can we waive the CodeQL?

**Review by pedroigor — APPROVED** [review]

**Review by mhajas — APPROVED** [review]
Thank you for the change @dweber019! I believe the test failures are unrelated. This looks good to me and it should be ready for merge.

Here is Jenkins run of springboot tests:
https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/175/

**Pull request merged**